### ISSUE TITLE
Custom Listener

### ISSUE THREAD

**Issue by CarlosUvaSilva** [rq/rq#1358]
(related to #1339 and #1337)

Is is possible to implement my own listener? I'm trying to listen to jobs posted from a Rails app but since RQ uses pickle and there's no documentation on using json, only that its available, I'm wondering if I can hack my own Redis listener

**Comment by selwin** [comment]
Yeah, better docs on using custom serializer is also something I'm looking to fix, but never found the time to.

Enqueueing a job using a JSON serializer looks like this (will add this to docs soon):
```python
queue = Queue(connection=redis, serializer=json)
queue.enqueue(say_hello)
```

More on the worker side here: https://python-rq.org/docs/workers/#worker-with-custom-serializer

**Comment by CarlosUvaSilva** [comment]
Hey @selwin thank you for replying

when using a custom serializer what fields are needed for the queu/job to work? I'm finding this very difficult/confusing :confused:

**Comment by selwin** [comment]
More work has been done on JSON serialization here: https://github.com/rq/rq/pull/1381

**Comment by lowercase00** [comment]
This is also something I would like to work on. Having language interoperability is big plus, specially since a lot of times `pickle` is not strictly necessary (in the sense that a JSONSerializer would work just fine). Celery does have a [protocol here](https://docs.celeryq.dev/en/stable/internals/protocol.html) and that has been very useful in the past (people have even implemented this protocol in Rust/Golang). 

`HMSET` with this payload on `rq:job:<job_id>` could probably work (it will save the job). The `data` is what's usually pickled if I'm not mistaken:

```json
{
    "created_at": "",
    "description": "",
    "enqueued_at": "",
    "last_heartbeat": "",
    "origin": "",
    "status": "",
    "success_callback_name": "",
    "failure_callback_name": "",
    "timeout": "",
    "worker_name": "",
    "data": {
        "_func_name": "",
        "_instance": "",
        "_args": "",
        "_kwargs": ""
    }
}
```

I guess all possible fields can be seen on the `to_dict()` method of the `job`, this is where the data that will be saved will be generated:
https://github.com/rq/rq/blob/c5a7a6b670582a20e353d5c5322304bf9ff1057c/rq/job.py#L682-L741

And then `RPUSH` with the `<job_id / UUID>` on the queue (`rq:queue:<job_id>`).
